### PR TITLE
[Ruby 2.5] Numerical comparison operators no longer hide exceptions from #coerce

### PR DESCRIPTION
**Pull request by andrykonchin**
https://bugs.ruby-lang.org/issues/7688

Related specs from ruby/ruby https://github.com/ruby/ruby/commit/bc1827e8825c558bcda14a214280caf83dc1215b#diff-715114ab5128543f66318d6fa6acffae

Related specs for `Bignum#<=>` https://github.com/ruby/spec/commit/f125a5d745748c340a0636222370b1ebdd412c7c

### CODE REVIEW

**Comment by andrykonchin** [comment]
Shared examples are very similar but I am not sure if it would be better to squash them in one general/parameterized example (actually two - for comparisions and arithmetic operations)

File: core/float/shared/arithmetic_exception_in_coerce.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
+require File.expand_path('../../fixtures/classes', __FILE__)
+
+describe :float_arithmetic_exception_in_coerce, shared: true do
+  ruby_version_is "" ... "2.5" do
+    it "rescues exception raised in other#coerce and raises TypeError" do
+      b = mock("numeric with failed #coerce")
+      b.should_receive(:coerce).and_raise(RuntimeError)
```
Comment (eregon) [inline core/float/shared/arithmetic_exception_in_coerce.rb]: Should this raise `FloatSpecs::CoerceError` for consistency?

File: core/float/shared/arithmetic_exception_in_coerce.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
+require File.expand_path('../../fixtures/classes', __FILE__)
+
+describe :float_arithmetic_exception_in_coerce, shared: true do
+  ruby_version_is "" ... "2.5" do
```
Comment (eregon) [inline core/float/shared/arithmetic_exception_in_coerce.rb]: The rest of the codebase does not use spaces around `...` (at least in version guards)

File: core/float/shared/arithmetic_exception_in_coerce.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    it "rescues exception raised in other#coerce and raises TypeError" do
```
Comment (eregon) [inline core/float/shared/arithmetic_exception_in_coerce.rb]: All exceptions or just StandardError?

**Comment by eregon** [comment]
> Shared examples are very similar but I am not sure if it would be better to squash them in one general/parameterized example (actually two - for comparisions and arithmetic operations)

Seems fine as you did, it's clearer and there is only minimal duplication.

File: core/float/shared/arithmetic_exception_in_coerce.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+      b = mock("numeric with failed #coerce")
+      b.should_receive(:coerce).and_raise(FloatSpecs::CoerceError)
+
+      # e.g. 1.0 > b
+      -> { 1.0.send(@method, b) }.should raise_error(TypeError, /MockObject can't be coerced into Float/)
+    end
+
+    it "does not rescue Exception and StandardError siblings raised in other#coerce" do
+      [ Exception.new, NoMemoryError.new, ScriptError.new, SecurityError.new,
+        SignalException.new('INT'), SystemExit.new, SystemStackError.new
```
Comment (eregon) [inline core/float/shared/arithmetic_exception_in_coerce.rb]: Seems fairly exhaustive :) I think 1 or 2 examples would be enough here, as we don't really want to test all the relations between these exception classes in this spec. Could you reduce it to 1 or 2 example classes?

**Comment by andrykonchin** [comment]
@eregon Done

**Comment by eregon** [comment]
Let's get this merged :) Thank you for adding proper specs for this!

**Pull request merged**